fms_extras/utils/cache/__init__.py

```diff
@@ -0,0 +1,330 @@
+    sequences that are being stored in the cache denoted by an integer ID.
+    """
+
+    @abc.abstractmethod
+    def allocate_tokens(
+        self,
+        num_tokens_per_sequence: List[int],
+        sequence_ids: Optional[List[int]] = None,
+    ) -> CacheDataWithMetadata:
+        """
+        allocate tokens in the kv-cache. If sequence ids are not given, this will be considered pre-fill for the prompt,
+        and sequence ids will be generated. If sequence ids are given, this will be considered allocating tokens for
+        generation and paged attention will be used
+
+        Parameters
+        ----------
+        num_tokens_per_sequence: List[int]
+            the number of tokens per sequence to expand in the kv-cache. This should correspond index-to-index with the
+            given sequence_ids
+        sequence_ids: List[int], optional
+            a list of sequence ids that will be expanded in the cache with generated tokens. If no sequence ids are
+            given, a new sequence id will be generated and allocation will be considered for the prompt
+            (default is None)
+
+        Returns
+        -------
+        CacheDataWithMetadata
+            a cache data object that includes metadata associated with it based on the current state of the
+            KVCacheManager for the given sequence ids.
+        """
+        pass
+
+    @abc.abstractmethod
+    def free_sequences(self, sequence_ids: List[int]):
+        """
+        free the given sequence ids from the kv-cache
+
+        Parameters
+        ----------
+        sequence_ids: List[int]
+            list of sequence ids to free
+        """
+        pass
+
+
+KVCache = Tuple[torch.Tensor, torch.Tensor]  # (key cache, value cache)
+
+
+@dataclasses.dataclass
+class OutOfPlaceCacheDataLayer(CacheDataLayer):
+    data_layer: Tuple[torch.Tensor, torch.Tensor]
+
+    def get_cache_type(self) -> str:
+        return "out-of-place"
+
+    def store(
+        self, keys: torch.Tensor, values: torch.Tensor
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        if self.data_layer is not None:
+            self.data_layer = (
+                torch.cat((self.data_layer[0], keys), dim=2),
+                torch.cat((self.data_layer[1], values), dim=2),
+            )
+            keys, values = self.data_layer
+        return keys, values
+
+    def is_filled(self) -> bool:
+        return self.data_layer is not None
+
+
+class OutOfPlaceCacheData(CacheData):
+    def __init__(self, data: List[Tuple[torch.Tensor, torch.Tensor]]):
+        self.data = data
+        self.max_sequence_length = (
+            0 if self.data[0] is None else self.data[0][0].size(2)
+        )
+
+    def get_layer(self, layer_index: int) -> OutOfPlaceCacheDataLayer:
+        return OutOfPlaceCacheDataLayer(data_layer=self.data[layer_index])
+
+    def is_filled(self) -> bool:
+        return self.data[0] is not None
+
+
+def flatten_batch(inp):
```

Comment:
this might need a better name/comment explaining that it's used for speculative decoding, it wasn't immediately obvious

Reply:
plus, what is a redundant token? what is sized as input/sized as output?

Reply:
This is the function that takes a tree of suffixes:
```
a b c
a b d
a e f
```
and returns `abcdef`. Tokens `a b` in the second line and token `a` in the third line are redundant because we can simply point the attention layer to the corresponding tokens in line 1, which will have the same values. 

But this probably does belong as a utility function in fms-extras